README.md
# update-nodejs-notifier [![npm](https://img.shields.io/npm/v/update-nodejs-notifier.svg?maxAge=2592000)](https://www.npmjs.com/package/update-nodejs-notifier) [![AppVeyor Status](https://ci.appveyor.com/api/projects/status/094e7unld01so0o7?svg=true)](https://ci.appveyor.com/project/jokeyrhyme/update-nodejs-notifier-js) [![Travis CI Status](https://travis-ci.org/jokeyrhyme/update-nodejs-notifier.js.svg?branch=master)](https://travis-ci.org/jokeyrhyme/update-nodejs-notifier.js)

tell your CLI users that their Node.js is old


## Usage

Set and forget.
Users will see a notification if their Node.js version needs updating.

This library will make no more than a single HTTP request every 5 days.


### Simple example

```js
const { updateNodejsNotifier } = require('update-nodejs-notifier')
updateNodejsNotifier()
```


### Comprehensive example

```js
const { updateNodejsNotifier } = require('update-nodejs-notifier')

updateNodejsNotifier({
  // listed in order of most-severe to least-severe
  // we display the first most-severe matching alert
  // these are the default settings
  notSupported: true, // alert if upstream Node.js support ended
  daysOld: Infinity, // alert if version released this long ago
  stableMajor: true, // alert if MAJOR version older than best "stable"
  stableMinor: false, // alert if MINOR version older than best "stable"
  stablePatch: false, // alert if PATCH version older than best "stable"
})
```


## See Also

-   [boxen-notify](https://github.com/jokeyrhyme/boxen-notify.js)

-   [hsipe](https://github.com/jokeyrhyme/hsipe.js)

-   [conf](https://github.com/sindresorhus/conf)

-   [package-engines-notifier](https://github.com/jokeyrhyme/package-engines-notifier.js)

-   [update-notifier](https://github.com/yeoman/update-notifier)

-   [nodejs-support-dates](https://github.com/jokeyrhyme/nodejs-support-dates.js)


CHANGELOG.md
# Change Log


## Unreleased


## 1.1.1 - 2017-02-01


### Changed

-   if caller passes `versions` in, don't fetch versions (better for testing)

-   allow caller to pass in `notify()` function (better for testing)


### Fixed

-   "update" message for unstable Node.js versions (#7)


## 1.1.0 - 2016-11-01


### Added

-   use [nodejs-support-dates](https://github.com/jokeyrhyme/nodejs-support-dates.js) to provide warnings about unsupported versions


## 1.0.1 - 2016-10-25


### Fixed

-   actually say that "Node.js" needs updating (oops)


test/bake.js
'use strict'

const test = require('ava')

const lib = require('../lib/bake.js')

test('exports an object', (t) => {
  t.is(typeof lib, 'object')
})

test('exports a "bake" function', (t) => {
  t.is(typeof lib.bake, 'function')
})

test('bake() returns a promised cake, with "versions" array', (t) => {
  return lib.bake({})
    .then((cake) => {
      t.truthy(cake && typeof cake === 'object')
      t.true(Array.isArray(cake.versions))
      t.true(cake.versions.length > 1)
      t.is(typeof cake.versions[0].version, 'string')
    })
})


test/index.js
'use strict'

const test = require('ava')

const lib = require('../index.js')

test('exports an object', (t) => {
  t.is(typeof lib, 'object')
})

test('exports an "updateNodejsNotifier" function', (t) => {
  t.is(typeof lib.updateNodejsNotifier, 'function')
})

// assumes current Node.js version is >=4
test('updateNodejsNotifier(), no matching alert', (t) => {
  const versions = [
    { version: 'v3.4.5', date: '', files: [], modules: '', lts: false },
    { version: 'v2.3.4', date: '', files: [], modules: '', lts: false },
    { version: 'v1.2.3', date: '', files: [], modules: '', lts: false }
  ]
  const notify = () => { notify.called = true }
  t.is(lib.updateNodejsNotifier({ notify, versions }), false)
  t.falsy(notify.called)
})

test('updateNodejsNotifier(), matching alert', (t) => {
  const versions = [
    { version: 'v3.4.5', date: '', files: [], modules: '', lts: false },
    { version: 'v2.3.4', date: '', files: [], modules: '', lts: false },
    { version: 'v1.2.3', date: '', files: [], modules: '', lts: false }
  ]
  const notify = (options) => {
    const message = options.message || ''
    t.truthy(message)
    notify.called = true
  }
  t.is(lib.updateNodejsNotifier({
    current: 'v1.2.3',
    notify,
    stableMajor: true,
    versions
  }), true)
  t.truthy(notify.called)
})

test('updateNodejsNotifier(), old unstable major version', (t) => {
  const TODAY = (new Date()).toISOString()
  const THREE_DAYS = 3 * 24 * 60 * 60 * 1000
  const OLD_DAY = (new Date(Date.now() - THREE_DAYS)).toISOString()
  const versions = [
    { version: 'v7.1.0', date: TODAY, files: [], modules: '', lts: false },
    { version: 'v6.9.0', date: TODAY, files: [], modules: '', lts: false },
    { version: 'v7.0.0', date: OLD_DAY, files: [], modules: '', lts: false }
  ]
  const notify = (options) => {
    const message = options.message || ''
    t.is(message, `Node.js update available v7.0.0 → v7.1.0
v7.0.0 is more than 1 days old`)
    notify.called = true
  }
  t.is(lib.updateNodejsNotifier({
    current: 'v7.0.0',
    daysOld: 1,
    notify,
    versions
  }), true)
  t.truthy(notify.called)
})


test/alerts.js
'use strict'

const test = require('ava')

const lib = require('../lib/alerts.js')

test('all alerts implement required interface', (t) => {
  t.true(lib.ALERTS && typeof lib.ALERTS === 'object')
  Object.keys(lib.ALERTS).forEach((key) => {
    const alert = lib.ALERTS[key]
    t.true(alert && typeof alert === 'object', `ALERTS[${key}]`)

    t.is(typeof alert.isEnabled, 'function', key, `ALERTS[${key}]`)
    t.is(typeof alert.isEnabled({}), 'boolean', `ALERTS[${key}]`)

    t.is(typeof alert.message, 'function', `ALERTS[${key}]`)
    t.is(typeof alert.message({}, ''), 'string', `ALERTS[${key}]`)

    t.is(typeof alert.test, 'function', `ALERTS[${key}]`)
  })
})

test('daysOld test()', (t) => {
  const TWO_DAYS = 2 * 24 * 60 * 60 * 1000
  const dayBeforeYesterday = (new Date(new Date() - TWO_DAYS)).toISOString()
  const list = [
    { version: 'v2.3.4', date: dayBeforeYesterday, files: [], modules: '', lts: false },
    { version: 'v1.2.3', date: '1970-01-01', files: [], modules: '', lts: false }
  ]
  t.is(
    lib.ALERTS['daysOld'].test({}, list, 'v1.2.3'),
    false,
    'default is Infinity, so an ancient date should raise no alert'
  )
  t.is(
    lib.ALERTS['daysOld'].test({ daysOld: 30 }, list, 'v1.2.3'),
    true,
    'ancient date is older than 30 days, alert!'
  )
  t.is(
    lib.ALERTS['daysOld'].test({ daysOld: 1 }, list, 'v2.3.4'),
    true,
    'day before yesterday is older than 1 day, alert!'
  )
})

test('matchAlerts() finding no alerts', (t) => {
  t.is(lib.matchAlerts({}, [], 'v1.2.3'), null)
})

test('matchAlerts() returns 1st matching alert', (t) => {
  const list = [
    { version: 'v3.4.5', date: '', files: [], modules: '', lts: false },
    { version: 'v2.3.4', date: '', files: [], modules: '', lts: false },
    { version: 'v1.2.3', date: '', files: [], modules: '', lts: false }
  ]
  const expected = lib.ALERTS['stableMajor']
  t.is(lib.matchAlerts({
    stableMajor: true,
    stableMinor: true
  }, list, 'v1.2.3'), expected)
})

test('notSupported test()', (t) => {
  const ltsFromDate = require('nodejs-support-dates').ltsFromDate

  const ltsMajor = ltsFromDate(new Date())
  t.is(
    lib.ALERTS['notSupported'].test({}, [], 'v5.0.0'),
    true,
    'v5.x is not supported now, alert!'
  )
  t.is(
    lib.ALERTS['notSupported'].test({}, [], `v${ltsMajor}.0.0`),
    false,
    'LTS versions are supported, no alert'
  )
})

test('stableMajor test()', (t) => {
  const list = [
    { version: 'v3.4.5', date: '', files: [], modules: '', lts: false },
    { version: 'v2.3.4', date: '', files: [], modules: '', lts: false },
    { version: 'v1.2.3', date: '', files: [], modules: '', lts: false }
  ]
  t.is(lib.ALERTS['stableMajor'].test({}, list, 'v1.2.3'), true)
  t.is(lib.ALERTS['stableMajor'].test({}, list, 'v2.2.0'), false)
})

test('stableMinor test()', (t) => {
  const list = [
    { version: 'v3.4.5', date: '', files: [], modules: '', lts: false },
    { version: 'v2.3.4', date: '', files: [], modules: '', lts: false },
    { version: 'v1.2.3', date: '', files: [], modules: '', lts: false }
  ]
  t.is(lib.ALERTS['stableMinor'].test({}, list, 'v2.2.0'), true)
  t.is(lib.ALERTS['stableMinor'].test({}, list, 'v2.3.1'), false)
})

test('stablePatch test()', (t) => {
  const list = [
    { version: 'v3.4.5', date: '', files: [], modules: '', lts: false },
    { version: 'v2.3.4', date: '', files: [], modules: '', lts: false },
    { version: 'v1.2.3', date: '', files: [], modules: '', lts: false }
  ]
  t.is(lib.ALERTS['stablePatch'].test({}, list, 'v2.3.1'), true)
  t.is(lib.ALERTS['stablePatch'].test({}, list, 'v2.3.4'), false)
})


lib/alerts.js
/* @flow */
'use strict'

const SemVer = require('semver')

const versions = require('./versions.js')

/* ::
import type { UpdateNodejsNotifierOptions } from '../index.js'
import type { NodejsVersion } from './versions.js'
export type Alert = {
  isEnabled: (options: UpdateNodejsNotifierOptions) => boolean;
  message: (
    options: UpdateNodejsNotifierOptions,
    currentVersion: string
  ) => string;
  test: (
    options: UpdateNodejsNotifierOptions,
    list: NodejsVersion[],
    currentVersion: string
  ) => boolean
}
*/

const NOT_SUPPORTED = 'notSupported'
const DAYS_OLD = 'daysOld'
const STABLE_MAJOR = 'stableMajor'
const STABLE_MINOR = 'stableMinor'
const STABLE_PATCH = 'stablePatch'

// "test" functions return true if the alert condition is met
// "test" function return false if there is nothing to worry about
const ALERTS = {}

ALERTS[NOT_SUPPORTED] = {
  isEnabled: (
    options /* : UpdateNodejsNotifierOptions */
  ) /* : boolean */ => {
    return !!options[NOT_SUPPORTED]
  },

  message: (
    options /* : UpdateNodejsNotifierOptions */,
    currentVersion /* : string */
  ) /* : string */ => {
    const chalk = require('chalk') // slow, so only do when needed
    return chalk.red(`${currentVersion} is no longer receiving security updates!`)
  },

  test: (
    options /* : UpdateNodejsNotifierOptions */,
    list /* : NodejsVersion[] */,
    currentVersion /* : string */
  ) /* : boolean */ => {
    const datesFromVersion = require('nodejs-support-dates').datesFromVersion

    const end = datesFromVersion(currentVersion).end
    if (end < Date.now()) {
      return true
    }

    return false
  }
}

ALERTS[DAYS_OLD] = {
  isEnabled: (
    options /* : UpdateNodejsNotifierOptions */
  ) /* : boolean */ => {
    return typeof options[DAYS_OLD] === 'number'
  },

  message: (
    options /* : UpdateNodejsNotifierOptions */,
    currentVersion /* : string */
  ) /* : string */ => {
    const chalk = require('chalk') // slow, so only do when needed
    const daysOld = '' + (options[DAYS_OLD] || Infinity)
    return chalk.yellow(`${currentVersion} is more than ${daysOld} days old`)
  },

  test: (
    options /* : UpdateNodejsNotifierOptions */,
    list /* : NodejsVersion[] */,
    currentVersion /* : string */
  ) /* : boolean */ => {
    const details = versions.find(list, currentVersion)
    const releaseDate = new Date(details.date)
    const msOld = (options.daysOld || Infinity) * 24 * 60 * 60 * 1000
    return (new Date()) - releaseDate > msOld
  }
}

ALERTS[STABLE_MAJOR] = {
  isEnabled: (
    options /* : UpdateNodejsNotifierOptions */
  ) /* : boolean */ => {
    return !!options[STABLE_MAJOR]
  },

  message: (
    options /* : UpdateNodejsNotifierOptions */,
    currentVersion /* : string */
  ) /* : string */ => {
    return `${currentVersion} is older than the latest stable version`
  },

  test: (
    options /* : UpdateNodejsNotifierOptions */,
    list /* : NodejsVersion[] */,
    currentVersion /* : string */
  ) /* : boolean */ => {
    const latestStable = versions.latestStableVersion(list)
    const latestStableMajor = SemVer.major(latestStable)
    return latestStableMajor > SemVer.major(currentVersion)
  }
}

ALERTS[STABLE_MINOR] = {
  isEnabled: (
    options /* : UpdateNodejsNotifierOptions */
  ) /* : boolean */ => {
    return !!options[STABLE_MINOR]
  },

  message: (
    options /* : UpdateNodejsNotifierOptions */,
    currentVersion /* : string */
  ) /* : string */ => {
    return `${currentVersion} is older than the latest stable version`
  },

  test: (
    options /* : UpdateNodejsNotifierOptions */,
    list /* : NodejsVersion[] */,
    currentVersion /* : string */
  ) /* : boolean */ => {
    const majorTest = ALERTS[STABLE_MAJOR].test
    if (majorTest(options, list, currentVersion)) {
      return true
    }
    const latestStable = versions.latestStableVersion(list)
    const latestStableMinor = SemVer.minor(latestStable)
    return latestStableMinor > SemVer.minor(currentVersion)
  }
}

ALERTS[STABLE_PATCH] = {
  isEnabled: (
    options /* : UpdateNodejsNotifierOptions */
  ) /* : boolean */ => {
    return !!options[STABLE_PATCH]
  },

  message: (
    options /* : UpdateNodejsNotifierOptions */,
    currentVersion /* : string */
  ) /* : string */ => {
    return `${currentVersion} is older than the latest stable version`
  },

  test: (
    options /* : UpdateNodejsNotifierOptions */,
    list /* : NodejsVersion[] */,
    currentVersion /* : string */
  ) /* : boolean */ => {
    const minorTest = ALERTS[STABLE_MINOR].test
    if (minorTest(options, list, currentVersion)) {
      return true
    }
    const latestStable = versions.latestStableVersion(list)
    const latestStablePatch = SemVer.patch(latestStable)
    return latestStablePatch > SemVer.patch(currentVersion)
  }
}

const SEVERITY_ORDER = [
  NOT_SUPPORTED,
  DAYS_OLD,
  STABLE_MAJOR,
  STABLE_MINOR,
  STABLE_PATCH
]

function matchAlerts (
  options /* : UpdateNodejsNotifierOptions */,
  list /* : NodejsVersion[] */,
  currentVersion /* : string */
) /* : ?Alert */ {
  // using .some() for early exit, unlike .forEach()
  let alert
  SEVERITY_ORDER.some((name) => {
    const candidate = ALERTS[name]
    if (candidate && candidate.isEnabled(options)) {
      try {
        if (candidate.test(options, list, currentVersion)) {
          alert = candidate
          return true
        }
      } catch (err) { /* ignore the error for now */ }
    }
  })
  return alert || null
}

module.exports = {
  NOT_SUPPORTED,
  DAYS_OLD,
  STABLE_MAJOR,
  STABLE_MINOR,
  STABLE_PATCH,

  ALERTS,
  SEVERITY_ORDER,

  matchAlerts
}


lib/versions.js
/* @flow */
'use strict'

const SemVer = require('semver')

/* ::
export type NodejsVersion = {
  version: string;
  date: string;
  files: string[];
  modules: string;
  lts: boolean | string;
};
*/

function find (
  list /* : NodejsVersion[] */,
  version /* : string */
) /* : NodejsVersion */ {
  for (let v = 0, vLength = list.length; v < vLength; v += 1) {
    const candidate = list[v]
    if (candidate.version === version) {
      return candidate
    }
  }
  throw new Error(`version ${version} not listed`)
}

function latestStableVersion (
  list /* : NodejsVersion[] */
) /* : string */ {
  // assume list is newest-to-oldest
  for (let v = 0, vLength = list.length; v < vLength; v += 1) {
    const version = list[v].version
    if (SemVer.major(version) % 2 === 0) {
      return version
    }
  }
  throw new Error('no stable versions identified')
}

function latestVersion (
  list /* : NodejsVersion[] */
) /* : string */ {
  // assume list is newest-to-oldest
  return list[0].version
}

module.exports = {
  find,
  latestStableVersion,
  latestVersion
}


index.js
/* @flow */
'use strict'

const path = require('path')

const getCake = require('hsipe').getCake
const putInOven = require('hsipe').putInOven
const semver = require('semver')

const NODEJS_URL = require('./lib/bake.js').NODEJS_URL

const cakeName = NODEJS_URL.replace(/\W+/g, '-')

const alerts = require('./lib/alerts.js')
const versions = require('./lib/versions.js')

/* ::
import type { NodejsVersion } from './lib/versions.js'
export type UpdateNodejsNotifierOptions = {
  notSupported?: boolean,
  daysOld?: number,
  stableMajor?: boolean,
  stableMinor?: boolean,
  stablePatch?: boolean,

  current?: string,
  notify?: ({ message: string }) => void,
  versions?: NodejsVersion[]
}
*/

const ONE_DAY = 7 * 24 * 60 * 60 * 1000

const defaults = {
  notSupported: true,
  daysOld: Infinity,
  stableMajor: true,
  stableMinor: false,
  stablePatch: false
}

function updateNodejsNotifier (
  options /* : ?UpdateNodejsNotifierOptions */
) /* : boolean */ {
  options = Object.assign({}, defaults, options)

  const nodejsVersions = options.versions || (() => {
    putInOven({
      bakePath: path.join(__dirname, 'lib', 'bake.js'),
      cakeName,
      interval: 5 * ONE_DAY
    })

    // try to continue on, in case we already started baking last time
    const cake = getCake({ cakeName })

    return cake.versions
  })()

  const current = options.current || process.version

  if (nodejsVersions) {
    const alert = alerts.matchAlerts(options, nodejsVersions, current)

    if (!alert) {
      return false
    }

    const chalk = require('chalk') // slow, so do this _after_ check

    let latest = versions.latestStableVersion(nodejsVersions)
    if (semver.gt(current, latest)) {
      latest = versions.latestVersion(nodejsVersions)
    }
    const message = chalk.blue('Node.js') +
      chalk.reset(' update available ') +
      chalk.dim(current) + chalk.reset(' → ') +
      chalk.green(latest) + chalk.reset('\n') +
      alert.message(options, current)

    // TODO: figure out better update instructions
    // ' \nRun ' + chalk.cyan('custom update command') + ' to update'

    if (typeof options.notify === 'function') {
      options.notify({ message })
    } else {
      require('boxen-notify').notify({ message })
    }
    return true
  }

  return false
}

module.exports = {
  updateNodejsNotifier
}
